datafusion/physical-plan/src/aggregates/group_values/column.rs

```diff
@@ -160,11 +162,112 @@ macro_rules! instantiate_primitive {
     };
 }
 
-fn append_col_value<C>(mut core: C, array: &ArrayRef, row: usize)
-where
-    C: FnMut(&ArrayRef, usize),
-{
-    core(array, row);
+struct AggregationHashTable<T: AggregationHashTableEntry> {
+    /// Raw table storing values in a `Vec`
+    raw_table: Vec<T>,
```

Comment:
Based on some experiments in changing hash join algorithm, I think it's likely `hashbrown` performs much better than implementing a hashtable ourselves although I would like to be surprised 🙂

Reply:
> Based on some experiments in changing hash join algorithm, I think it's likely `hashbrown` performs much better than implementing a hashtable ourselves although I would like to be surprised 🙂

🤔 Even if we can perform something like `vectorized compare` or `vectorized append` in our hashtable?

I found in `multi group by` case, we will perform the `compare` for each row leading to the array downcasting again and again... And actually the `downcast` operation will be compiled to many asm codes....

And I foudn we can't eliminate it and perform the `vectorized compare` with `hashbrown`...

```
    fn equal_to_inner(&self, lhs_row: usize, array: &ArrayRef, rhs_row: usize) -> bool {
        let array = array.as_byte_view::<B>();
```

Reply:
We can still do "vectorized compare" by doing the lookup in the hashtable (based on hash value only) and the vectorized equality check separately. That way you still can use the fast hashtable, but move the equality check to a separate/vectorized step.
That's at least what is done in the vectorized join implementation :). I changed it before to use a `Vec`-based index like you did here, but that performed significantly worse.

Reply:
The reason I think is that the lookup is incredibly well optimized using the swiss table design and you get fewer 'false" candidates to check for, while we can still use the vectorized/type specialized equality check.

Reply:
Make sense, thank you!